### ISSUE TITLE
[feature request]: system.text.json JsonExtensionData support for internal and private/protected

### ISSUE THREAD

**Issue by drdamour** [dotnet/runtime#78775]
I like to use JsonExtensionData to detect if the producer starts adding things to the “extensible” json that they produce that my serializer types aren't handling.  I hook these up to log traps that let my program tell my team we should consider changes.

I also like my types to be immutable on their public surface (everywhere if i can have it).  Hoping for #31645 for that

And i prefer to have my dictionary hidden from
public surface since i dont want consumers to use stuff that hasnt really gone through vetting.

but it seems system.text.json cant leverage a non public prop for JsonExtensionData cause nothing can per #31511 but i propose JsonExtensionData is a special case that should be considered since the workarounds in that ticket dont seem to apply.

**Comment by eiriktsarpalis** [comment]
We might consider extending `JsonIncludeAttribute` so that it can be used with non-public members, including extension properties.

In the meantime, it should be possible to work around your issue in .NET 7 using a custom contract resolver:
```C#
var options = new JsonSerializerOptions
{
    TypeInfoResolver = new DefaultJsonTypeInfoResolver()
    {
        Modifiers =
        {
            static typeInfo =>
            {
                if (typeInfo.Kind == JsonTypeInfoKind.Object && 
                    typeInfo.Properties.All(prop => !prop.IsExtensionData))
                {
                    PropertyInfo? extensionProp = typeInfo.Type
                        .GetProperties(BindingFlags.Instance | BindingFlags.NonPublic)
                        .FirstOrDefault(prop => prop.GetCustomAttribute<JsonExtensionDataAttribute>() != null);

                    if (extensionProp != null)
                    {
                        JsonPropertyInfo jsonPropertyInfo = typeInfo.CreateJsonPropertyInfo(extensionProp.PropertyType, extensionProp.Name);
                        jsonPropertyInfo.Get = extensionProp.GetValue;
                        jsonPropertyInfo.Set = extensionProp.SetValue;
                        jsonPropertyInfo.IsExtensionData = true;
                        typeInfo.Properties.Add(jsonPropertyInfo);
                    }
                }
            }
        }
    }
};

MyPoco result = JsonSerializer.Deserialize<MyPoco>("""{"foo":"bar"}""", options);
Console.WriteLine(result.Data["foo"]); // "bar"

public class MyPoco
{
    [JsonExtensionData]
    internal Dictionary<string, object>? Data { get; set; }
}
```

**Comment by drdamour** [comment]
that's awesome, i didn't catch the IsExtensionData of the contract stuff.

**Comment by l-404-l** [comment]
> We might consider extending `JsonIncludeAttribute` so that it can be used with non-public members, including extension properties.
> 
> In the meantime, it should be possible to work around your issue in .NET 7 using a custom contract resolver:
> 
> ```cs
> var options = new JsonSerializerOptions
> {
>     TypeInfoResolver = new DefaultJsonTypeInfoResolver()
>     {
>         Modifiers =
>         {
>             static typeInfo =>
>             {
>                 if (typeInfo.Kind == JsonTypeInfoKind.Object && 
>                     typeInfo.Properties.All(prop => !prop.IsExtensionData))
>                 {
>                     PropertyInfo? extensionProp = typeInfo.Type
>                         .GetProperties(BindingFlags.Instance | BindingFlags.NonPublic)
>                         .FirstOrDefault(prop => prop.GetCustomAttribute<JsonExtensionDataAttribute>() != null);
> 
>                     if (extensionProp != null)
>                     {
>                         JsonPropertyInfo jsonPropertyInfo = typeInfo.CreateJsonPropertyInfo(extensionProp.PropertyType, extensionProp.Name);
>                         jsonPropertyInfo.Get = extensionProp.GetValue;
>                         jsonPropertyInfo.Set = extensionProp.SetValue;
>                         jsonPropertyInfo.IsExtensionData = true;
>                         typeInfo.Properties.Add(jsonPropertyInfo);
>                     }
>                 }
>             }
>         }
>     }
> };
> 
> MyPoco result = JsonSerializer.Deserialize<MyPoco>("""{"foo":"bar"}""", options);
> Console.WriteLine(result.Data["foo"]); // "bar"
> 
> public class MyPoco
> {
>     [JsonExtensionData]
>     internal Dictionary<string, object>? Data { get; set; }
> }
> ```

Yes please extend JsonIncludes to non-public members would help when creating complex models.
